### PR TITLE
Make Kotlin Entities Immutable

### PR DESCRIPTION
**Pull request by SHeimlich**

### CODE REVIEW

File: src/tools/schema2kotlin.ts [hunk 1]
```diff
@@ -101,11 +101,13 @@ class KotlinGenerator implements ClassGenerator {
   fields: string[] = [];
   fieldVals: string[] = [];
   setFields: string[] = [];
-  fieldSets: string[] = [];
   encode: string[] = [];
   decode: string[] = [];
   fieldsReset: string[] = [];
   getUnsetFields: string[] = [];
+  fieldsForCopyDecl: string[] = [];
```
Comment (alxrsngrtn) [inline src/tools/schema2kotlin.ts]: Either in this PR or a near future one: [Destructuring declarations](https://kotlinlang.org/docs/reference/multi-declarations.html) (`component1`, `component2`, etc.) should also be implemented. At least, we should do this if we want to be more in like with [Kotlin data class](https://kotlinlang.org/docs/reference/data-classes.html#data-classes-and-destructuring-declarations) expectations.

File: src/tools/tests/goldens/generated-schemas.jvm.kt [hunk 2]
```diff
@@ -29,19 +27,16 @@ class GoldInternal1() : Entity {
         this.val_ = val_
     }
 
-    override fun isSet(): Boolean {
-        return _val_Set
+    fun copy(
+        val_: String = this.val_
+    ) : GoldInternal1 {
+      return GoldInternal1(
```
Comment (csilvestrini) [inline src/tools/tests/goldens/generated-schemas.jvm.kt]: Can you fix the indentation here in the code generator?

File: src/tools/tests/goldens/generated-schemas.jvm.kt [hunk 3]
```diff
@@ -104,28 +91,25 @@ class Gold_Data() : Entity {
...
 
-    override fun isSet(): Boolean {
-        return _numSet && _txtSet && _lnkSet && _flgSet
+    fun copy(
+        num: Double = this.num,
+        txt: String = this.txt,
+        lnk: String = this.lnk,
+        flg: Boolean = this.flg
+    ) : Gold_Data {
+      return Gold_Data(
```
Comment (csilvestrini) [inline src/tools/tests/goldens/generated-schemas.jvm.kt]: I feel stupid nitpicking style for a generated file... But I'm going to do it anyway :stuck_out_tongue: 

```
fun copy(
    // args...
) = Gold_Data(
    // args...
)
```

File: javatests/arcs/sdk/wasm/CollectionApiTest.kt [hunk 4]
```diff
@@ -31,26 +33,28 @@ class CollectionApiTest : AbstractCollectionApiTest() {
...
-                d3.flg = !iter.hasNext()
-                outHandle.store(d3)
+                outHandle.store(d1.copy(
+                    txt = "ne",
+                    flg = !iter.hasNext()
+                ))
             }
             "case5" -> {
-                val extra = CollectionApiTest_IoHandle()
+                val extra = CollectionApiTest_IoHandle().copy(txt = "abc")
```
[the remark this answers is not in the document]
Comment (csilvestrini) [reply javatests/arcs/sdk/wasm/CollectionApiTest.kt]: Constructor already exists it turns out

**Pull request merged**